In [128]:
import pandas as pd
import numpy as np
import requests
from datetime import datetime
from bs4 import BeautifulSoup
from utils import read_config, MySQLAgent
import time, random

In [129]:
config = read_config('.env/local_conn.json')
sql_agent = MySQLAgent(config['local_mysql'])

In [130]:
query = """
show tables
"""

sql_agent.read_table(query=query)

tables_in_generaldb
0        balancesheet
1       port_activity
2       profitandlose

### Helper functions

In [131]:
def get_acct_data_and_rename(df, row_name, new_name):

    result = df[df['acct_name'] == row_name][['year_season', 'this_year_amt']].reset_index(drop=True).rename(columns={'this_year_amt':new_name})
    result[new_name] = result[new_name].str.replace(',', '').astype(int)

    return result

In [132]:
query = """
select *
from balancesheet
where stock = '2330' # and season = 4
"""

df_balance = sql_agent.read_table(query=query)

In [133]:
df_balance

acct_name  this_year_amt this_year_percent  \
0                          流動資產                                    
1                       現金及約當現金    645,670,527             29.52   
2           透過損益按公允價值衡量之金融資產－流動      3,084,399              0.14   
3       透過其他綜合損益按公允價值衡量之金融資產－流動    107,313,205              4.91   
4              按攤銷後成本衡量之金融資產－流動      4,179,386              0.19   
...                         ...            ...               ...   
1393                       權益總額  3,665,716,254             63.33   
1394                    負債及權益總計  5,787,891,082            100.00   
1395              待註銷股本股數（單位：股）      1,401,750                     
1396    預收股款（權益項下）之約當發行股數（單位：股）              0                     
1397  母公司暨子公司所持有之母公司庫藏股股數（單位：股）              0                     

      last_year_amt last_year_percent   report_name report_time_raw  \
0                                      BalanceSheet       民國108年第1季   
1       577,782,963             28.20  BalanceSheet       民國108年第1季   
2           963,915              0.05  BalanceSheet       民國108年第1季   
3        95,713,446              4.67  BalanceSheet       民國108年第1季   
4         9,888,741              0.48  BalanceSheet       民國108年第1季   
...             ...               ...           ...             ...   
1393  3,092,897,598             61.30  BalanceSheet       民國113年第1季   
1394  5,045,844,348            100.00  BalanceSheet       民國113年第1季   
1395        419,466                    BalanceSheet       民國113年第1季   
1396              0                    BalanceSheet       民國113年第1季   
1397              0                    BalanceSheet       民國113年第1季   

      period_year  season creation_date year_season stock  
0             108       1    2024-09-07       108-1  2330  
1             108       1    2024-09-07       108-1  2330  
2             108       1    2024-09-07       108-1  2330  
3             108       1    2024-09-07       108-1  2330  
4             108       1    2024-09-07       108-1  2330  
...           ...     ...           ...         ...   ...  
1393          113       1    2024-09-07       113-1  2330  
1394          113       1    2024-09-07       113-1  2330  
1395          113       1    2024-09-07       113-1  2330  
1396          113       1    2024-09-07       113-1  2330  
1397          113       1    2024-09-07       113-1  2330  

[1398 rows x 12 columns]

In [134]:
df_balance.acct_name.unique()

array(['流動資產', '現金及約當現金', '透過損益按公允價值衡量之金融資產－流動',
       '透過其他綜合損益按公允價值衡量之金融資產－流動', '按攤銷後成本衡量之金融資產－流動', '避險之金融資產－流動',
       '應收帳款淨額', '應收帳款－關係人淨額', '其他應收款－關係人淨額', '存貨', '其他流動資產', '流動資產合計',
       '非流動資產', '透過其他綜合損益按公允價值衡量之金融資產－非流動', '按攤銷後成本衡量之金融資產－非流動',
       '採用權益法之投資', '不動產、廠房及設備', '使用權資產', '無形資產', '遞延所得稅資產', '其他非流動資產',
       '非流動資產合計', '資產總額', '流動負債', '短期借款', '透過損益按公允價值衡量之金融負債－流動',
       '避險之金融負債－流動', '應付帳款', '應付帳款－關係人', '其他應付款', '本期所得稅負債', '其他流動負債',
       '流動負債合計', '非流動負債', '應付公司債', '遞延所得稅負債', '租賃負債－非流動', '其他非流動負債',
       '非流動負債合計', '負債總額', '歸屬於母公司業主之權益', '股本', '普通股股本', '股本合計', '資本公積',
       '資本公積－發行溢價', '資本公積-認列對子公司所有權權益變動數', '資本公積－受贈資產',
       '資本公積－採用權益法認列關聯企業及合資股權淨值之變動數', '資本公積－合併溢額', '資本公積合計', '保留盈餘',
       '法定盈餘公積', '特別盈餘公積', '未分配盈餘（或待彌補虧損）', '保留盈餘合計', '其他權益', '其他權益合計',
       '歸屬於母公司業主之權益合計', '非控制權益', '權益總額', '負債及權益總計',
       '預收股款（權益項下）之約當發行股數（單位：股）', '母公司暨子公司所持有之母公司庫藏股股數（單位：股）', '應付短期票券',
       '長期借款', '資本公積－實際取得或處分子公司股權價格與帳面價值差額', '資本公積－限制員工權利股票', '庫藏股票',
      

In [135]:
query = """
select *
from profitandlose
where stock = '2330' # and season = 4
"""

df_profitloss = sql_agent.read_table(query=query)

In [136]:
df_profitloss.iloc[:25]

acct_name this_year_amt this_year_percent  \
0                          營業收入合計   218,704,469            100.00   
1                          營業成本合計   128,352,344             58.69   
2                        營業毛利（毛損）    90,352,125             41.31   
3                       已實現銷貨（損）益         5,432              0.00   
4                      營業毛利（毛損）淨額    90,357,557             41.31   
5                            營業費用                                   
6                            推銷費用     1,459,973              0.67   
7                            管理費用     4,140,729              1.89   
8                          研究發展費用    20,417,311              9.34   
9                          營業費用合計    26,018,013             11.90   
10                      其他收益及費損淨額                                   
11                      其他收益及費損淨額       -73,521             -0.03   
12                       營業利益（損失）    64,266,023             29.38   
13                       營業外收入及支出                                   
14                           其他收入     4,408,776              2.02   
15                      其他利益及損失淨額       -27,573             -0.01   
16                         財務成本淨額       899,065              0.41   
17         採用權益法認列之關聯企業及合資損益之份額淨額       433,491              0.20   
18                     營業外收入及支出合計     3,915,629              1.79   
19                       稅前淨利（淨損）    68,181,652             31.18   
20                    所得稅費用（利益）合計     6,794,342              3.11   
21                 繼續營業單位本期淨利（淨損）    61,387,310             28.07   
22                       本期淨利（淨損）    61,387,310             28.07   
23                     其他綜合損益(淨額)                                   
24  透過其他綜合損益按公允價值衡量之權益工具投資未實現評價損益        53,541              0.02   

   last_year_amt last_year_percent    report_name report_time_raw  \
0    248,078,671            100.00  ProfitAndLose       民國108年第1季   
1    123,103,977             49.62  ProfitAndLose       民國108年第1季   
2    124,974,694             50.38  ProfitAndLose       民國108年第1季   
3       -117,155             -0.05  ProfitAndLose       民國108年第1季   
4    124,857,539             50.33  ProfitAndLose       民國108年第1季   
5                                   ProfitAndLose       民國108年第1季   
6      1,448,092              0.58  ProfitAndLose       民國108年第1季   
7      4,851,708              1.96  ProfitAndLose       民國108年第1季   
8     20,428,594              8.23  ProfitAndLose       民國108年第1季   
9     26,728,394             10.77  ProfitAndLose       民國108年第1季   
10                                  ProfitAndLose       民國108年第1季   
11    -1,302,199             -0.52  ProfitAndLose       民國108年第1季   
12    96,826,946             39.03  ProfitAndLose       民國108年第1季   
13                                  ProfitAndLose       民國108年第1季   
14     3,154,642              1.27  ProfitAndLose       民國108年第1季   
15        88,208              0.04  ProfitAndLose       民國108年第1季   
16       807,966              0.33  ProfitAndLose       民國108年第1季   
17       681,791              0.27  ProfitAndLose       民國108年第1季   
18     3,116,675              1.26  ProfitAndLose       民國108年第1季   
19    99,943,621             40.29  ProfitAndLose       民國108年第1季   
20    10,156,047              4.09  ProfitAndLose       民國108年第1季   
21    89,787,574             36.19  ProfitAndLose       民國108年第1季   
22    89,787,574             36.19  ProfitAndLose       民國108年第1季   
23                                  ProfitAndLose       民國108年第1季   
24       -19,507             -0.01  ProfitAndLose       民國108年第1季   

    period_year season creation_date year_season stock  
0           108      1    2024-09-07       108-1  2330  
1           108      1    2024-09-07       108-1  2330  
2           108      1    2024-09-07       108-1  2330  
3           108      1    2024-09-07       108-1  2330  
4           108      1    2024-09-07       108-1  2330  
5           108      1    2024-09-07       108-1  2330  
6           108      1    2024-09-0

In [137]:
df_profitloss[(df_profitloss['year_season'] == '112-4')].iloc[:25]

acct_name this_year_amt this_year_percent  \
1108                                營業收入合計     625528856              None   
1109                                營業成本合計     293760931              None   
1110                              營業毛利（毛損）     331767925              None   
1111                            營業毛利（毛損）淨額     331767925              None   
1112                                  推銷費用       2975547              None   
1113                                  管理費用      18235380              None   
1114                                研究發展費用      50410094              None   
1115                                營業費用合計      71621021              None   
1116                             其他收益及費損淨額         57599              None   
1117                              營業利益（損失）     260204503              None   
1118                                  利息收入      17408788              None   
1119                                  其他收入         76257              None   
1120                             其他利益及損失淨額       2194914              None   
1121                                財務成本淨額       2918597              None   
1122                採用權益法認列之關聯企業及合資損益之份額淨額       1315124              None   
1123                            營業外收入及支出合計      18076486              None   
1124                              稅前淨利（淨損）     278280989              None   
1125                           所得稅費用（利益）合計      39974788              None   
1126                        繼續營業單位本期淨利（淨損）     238306201              None   
1127                              本期淨利（淨損）     238306201              None   
1128         透過其他綜合損益按公允價值衡量之權益工具投資未實現評價損益        966670              None   
1129                       避險工具之損益－不重分類至損益             0              None   
1130  採用權益法認列之關聯企業及合資之其他綜合損益之份額-不重分類至損益之項目        -20402              None   
1131                        與不重分類之項目相關之所得稅       -124671              None   
1132                           不重分類至損益之項目：        447583              None   

     last_year_amt last_year_percent report_name report_time_raw  period_year  \
1108          None              None        None            None          112   
1109          None              None        None            None          112   
1110          None              None        None            None          112   
1111          None              None        None            None          112   
1112          None              None        None            None          112   
1113          None              None        None            None          112   
1114          None              None        None            None          112   
1115          None              None        None            None          112   
1116          None              None        None            None          112   
1117          None              None        None            None          112   
1118          None              None        None            None          112   
1119          None              None        None            None          112   
1120          None              None        None            None          112   
1121          None              None        None            None          112   
1122          None              None        None            None          112   
1123          None              None        None            None          112   
1124          None              None        None            None          112   
1125          None              None        None            None          112   
1126          None              None        None            None          112   
1127          None              None        None            None          112   
1128          None              None        None            None          112   
1129          None              None        None            None          112   
1130          None              None        None            None          112   
1131          None              None        None            

# Financial Structure

### Debut Burden Ratio(DBR) (資產負債比)

In [138]:
# 資產總額
df_total_assets = get_acct_data_and_rename(df_balance, '資產總額', 'total_assets')
df_total_assets['total_assets_varation'] = df_total_assets['total_assets'].pct_change()*100
df_total_assets

year_season  total_assets  total_assets_varation
0        108-1    2187436785                    NaN
1        108-2    2239343671               2.372955
2        108-3    2134234450              -4.693751
3        108-4    2264805032               6.117912
4        109-1    2343295696               3.465670
5        109-2    2448664522               4.496608
6        109-3    2635572214               7.633046
7        109-4    2760711405               4.748084
8        110-1    2919131461               5.738378
9        110-2    3092815189               5.949843
10       110-3    3332311884               7.743647
11       110-4    3725503455              11.799363
12       111-1    3992676667               7.171466
13       111-2    4345941335               8.847816
14       111-3    4643301766               6.842256
15       111-4    4964778878               6.923459
16       112-1    5045844348               1.632811
17       112-2    5149465045               2.053585
18       112-3    5484556381               6.507304
19       112-4    5532371215               0.871809
20       113-1    5787891082               4.618632

In [139]:
# 負債總額
df_total_liability = get_acct_data_and_rename(df_balance, '負債總額', 'total_liabilities')
df_total_liability['total_liabilities_varation'] = df_total_liability['total_liabilities'].pct_change()*100
df_total_liability

year_season  total_liabilities  total_liabilities_varation
0        108-1          443919183                         NaN
1        108-2          684922144                   54.289828
2        108-3          546737449                  -20.175241
3        108-4          642709606                   17.553610
4        109-1          666267165                    3.665350
5        109-2          727451140                    9.183099
6        109-3          847305839                   16.475979
7        109-4          910089406                    7.409788
8        110-1          978134499                    7.476748
9        110-2         1098844616                   12.340851
10       110-3         1253991574                   14.119099
11       110-4         1554770250                   23.985702
12       111-1         1671207913                    7.489059
13       111-2         1835479210                    9.829495
14       111-3         1890985607                    3.024082
15       111-4         2004290011                    5.991817
16       112-1         1952946750                   -2.561668
17       112-2         1943996846                   -0.458277
18       112-3         2111738274                    8.628688
19       112-4         2049108368                   -2.965799
20       113-1         2122174828                    3.565768

In [140]:
# DBR: Debut Burden Ratio
df_total_assets_liability = pd.merge(df_total_assets, df_total_liability, how='inner', on='year_season')
df_total_assets_liability['DBR'] = (round(df_total_assets_liability['total_liabilities'].astype(int) / df_total_assets_liability['total_assets'].astype(int) * 100, 2)).astype(str) + '%'
df_total_assets_liability

year_season  total_assets  total_assets_varation  total_liabilities  \
0        108-1    2187436785                    NaN          443919183   
1        108-2    2239343671               2.372955          684922144   
2        108-3    2134234450              -4.693751          546737449   
3        108-4    2264805032               6.117912          642709606   
4        109-1    2343295696               3.465670          666267165   
5        109-2    2448664522               4.496608          727451140   
6        109-3    2635572214               7.633046          847305839   
7        109-4    2760711405               4.748084          910089406   
8        110-1    2919131461               5.738378          978134499   
9        110-2    3092815189               5.949843         1098844616   
10       110-3    3332311884               7.743647         1253991574   
11       110-4    3725503455              11.799363         1554770250   
12       111-1    3992676667               7.171466         1671207913   
13       111-2    4345941335               8.847816         1835479210   
14       111-3    4643301766               6.842256         1890985607   
15       111-4    4964778878               6.923459         2004290011   
16       112-1    5045844348               1.632811         1952946750   
17       112-2    5149465045               2.053585         1943996846   
18       112-3    5484556381               6.507304         2111738274   
19       112-4    5532371215               0.871809         2049108368   
20       113-1    5787891082               4.618632         2122174828   

    total_liabilities_varation     DBR  
0                          NaN  20.29%  
1                    54.289828  30.59%  
2                   -20.175241  25.62%  
3                    17.553610  28.38%  
4                     3.665350  28.43%  
5                     9.183099  29.71%  
6                    16.475979  32.15%  
7                     7.409788  32.97%  
8                     7.476748  33.51%  
9                    12.340851  35.53%  
10                   14.119099  37.63%  
11                   23.985702  41.73%  
12                    7.489059  41.86%  
13                    9.829495  42.23%  
14                    3.024082  40.73%  
15                    5.991817  40.37%  
16                   -2.561668   38.7%  
17                   -0.458277  37.75%  
18                    8.628688   38.5%  
19                   -2.965799  37.04%  
20                    3.565768  36.67%

- 108年~110年，債務比例增加，可能是要擴廠或增加設備等等

### Long term funds to fixed assets ratio(長期資金佔固定資產比率)

(股東權益淨額 + 長期負債) / 固定資產淨額

在評估企業財務穩健性時，長期資金佔固定資產比率是一個重要的指標。長期資金，如股本和長期負債，被視為相對穩定的資金來源，而固定資產是指長期持有、無法迅速變現的資產，如廠房和設備。長期資金佔固定資產比率衡量長期資金對固定資產的覆蓋程度，反映企業償還固定資產投資的能力。

一般而言，較高的長期資金佔固定資產比率表示企業的財務結構較為穩健，因為它意味著企業有更多資金可供償還固定資產投資。然而，過高的比率也並非好事。當長期資金過多時，企業可能會無法有效利用資金，導致資金利用率低下，進而影響盈利能力。因此，在評估企業的財務穩健性時，投資者需要綜合考量長期資金佔固定資產比率和其他財務指標，例如負債對股東權益比率和現金流量狀況，才能做出全面判斷。

In [141]:
# 固定資產: 不動產、廠房及設備
df_fixed_assets = get_acct_data_and_rename(df_balance, '不動產、廠房及設備', 'fixed_assets')
df_fixed_assets

year_season  fixed_assets
0        108-1    1107651816
1        108-2    1142871184
2        108-3    1197955298
3        108-4    1352377405
4        109-1    1438215285
5        109-2    1493560950
6        109-3    1504438119
7        109-4    1555589120
8        110-1    1658551290
9        110-2    1722679133
10       110-3    1828894565
11       110-4    1975118704
12       111-1    2104330654
13       111-2    2248318787
14       111-3    2411490251
15       111-4    2693836970
16       112-1    2833396892
17       112-2    2947233393
18       112-3    3132664367
19       112-4    3064474984
20       113-1    3051847989

In [142]:
# 長期借款
df_long_term_liabilities = get_acct_data_and_rename(df_balance, '長期借款', 'long-term_liabilities')
df_long_term_liabilities

year_season  long-term_liabilities
0        109-3                 982449
1        109-4                1967611
2        110-1                2623824
3        110-2                3171331
4        110-3                3472190
5        110-4                3309131
6        111-1                3087714
7        111-2                3318773
8        111-3                3055178
9        111-4                4760047
10       112-1                6147668
11       112-2                5611138
12       112-3                5013050
13       112-4                4382965
14       113-1               10020132

In [143]:
df_balance[df_balance['acct_name'] == '長期借款']

acct_name this_year_amt this_year_percent last_year_amt  \
421       長期借款       982,449              0.04             0   
486       長期借款     1,967,611              0.07             0   
552       長期借款     2,623,824              0.09             0   
619       長期借款     3,171,331              0.10             0   
685       長期借款     3,472,190              0.10       982,449   
751       長期借款     3,309,131              0.09     1,967,611   
817       長期借款     3,087,714              0.08     2,623,824   
885       長期借款     3,318,773              0.08     3,171,331   
953       長期借款     3,055,178              0.07     3,472,190   
1021      長期借款     4,760,047              0.10     3,309,131   
1089      長期借款     6,147,668              0.12     3,087,714   
1160      長期借款     5,611,138              0.11     3,318,773   
1228      長期借款     5,013,050              0.09     3,055,178   
1296      長期借款     4,382,965              0.08     4,760,047   
1364      長期借款    10,020,132              0.17     6,147,668   

     last_year_percent   report_name report_time_raw  period_year  season  \
421               0.00  BalanceSheet       民國109年第3季          109       3   
486               0.00  BalanceSheet       民國109年第4季          109       4   
552               0.00  BalanceSheet       民國110年第1季          110       1   
619               0.00  BalanceSheet       民國110年第2季          110       2   
685               0.04  BalanceSheet       民國110年第3季          110       3   
751               0.07  BalanceSheet       民國110年第4季          110       4   
817               0.09  BalanceSheet       民國111年第1季          111       1   
885               0.10  BalanceSheet       民國111年第2季          111       2   
953               0.10  BalanceSheet       民國111年第3季          111       3   
1021              0.09  BalanceSheet       民國111年第4季          111       4   
1089              0.08  BalanceSheet       民國112年第1季          112       1   
1160              0.08  BalanceSheet       民國112年第2季          112       2   
1228              0.07  BalanceSheet       民國112年第3季          112       3   
1296              0.10  BalanceSheet       民國112年第4季          112       4   
1364              0.12  BalanceSheet       民國113年第1季          113       1   

     creation_date year_season stock  
421     2024-09-07       109-3  2330  
486     2024-09-07       109-4  2330  
552     2024-09-07       110-1  2330  
619     2024-09-07       110-2  2330  
685     2024-09-07       110-3  2330  
751     2024-09-07       110-4  2330  
817     2024-09-07       111-1  2330  
885     2024-09-07       111-2  2330  
953     2024-09-07       111-3  2330  
1021    2024-09-07       111-4  2330  
1089    2024-09-07       112-1  2330  
1160    2024-09-07       112-2  2330  
1228    2024-09-07       112-3  2330  
1296    2024-09-07       112-4  2330  
1364    2024-09-07       113-1  2330

In [144]:
# 股東權益＝資產－負債
df_shareholder_equity = df_total_assets_liability[['year_season', 'total_assets', 'total_liabilities']].copy()
df_shareholder_equity['shareholder_equity'] = df_shareholder_equity['total_assets'] - df_shareholder_equity['total_liabilities']
df_shareholder_equity

year_season  total_assets  total_liabilities  shareholder_equity
0        108-1    2187436785          443919183          1743517602
1        108-2    2239343671          684922144          1554421527
2        108-3    2134234450          546737449          1587497001
3        108-4    2264805032          642709606          1622095426
4        109-1    2343295696          666267165          1677028531
5        109-2    2448664522          727451140          1721213382
6        109-3    2635572214          847305839          1788266375
7        109-4    2760711405          910089406          1850621999
8        110-1    2919131461          978134499          1940996962
9        110-2    3092815189         1098844616          1993970573
10       110-3    3332311884         1253991574          2078320310
11       110-4    3725503455         1554770250          2170733205
12       111-1    3992676667         1671207913          2321468754
13       111-2    4345941335         1835479210          2510462125
14       111-3    4643301766         1890985607          2752316159
15       111-4    4964778878         2004290011          2960488867
16       112-1    5045844348         1952946750          3092897598
17       112-2    5149465045         1943996846          3205468199
18       112-3    5484556381         2111738274          3372818107
19       112-4    5532371215         2049108368          3483262847
20       113-1    5787891082         2122174828          3665716254

In [145]:
parent_table = df_balance[['year_season']].drop_duplicates()
df_temp1 = pd.merge(df_fixed_assets, df_long_term_liabilities, how='left', on='year_season')
df_temp2 = pd.merge(df_temp1, df_shareholder_equity, how='left', on='year_season')
# in here, we should check the last amt in the original table. 0 here is the result after checking
# create a check function laterly
df_longterm_funds_to_fixed_assets_retio = pd.merge(parent_table, df_temp2, how='left', on='year_season').fillna(0)
df_longterm_funds_to_fixed_assets_retio['longterm_funds_to_fixed_assets_retio'] = round((df_longterm_funds_to_fixed_assets_retio['long-term_liabilities'].astype(int) + df_longterm_funds_to_fixed_assets_retio['shareholder_equity'])/ df_longterm_funds_to_fixed_assets_retio['fixed_assets'].astype(int), 2)
df_longterm_funds_to_fixed_assets_retio

year_season  fixed_assets  long-term_liabilities  total_assets  \
0        108-1    1107651816                    0.0    2187436785   
1        108-2    1142871184                    0.0    2239343671   
2        108-3    1197955298                    0.0    2134234450   
3        108-4    1352377405                    0.0    2264805032   
4        109-1    1438215285                    0.0    2343295696   
5        109-2    1493560950                    0.0    2448664522   
6        109-3    1504438119               982449.0    2635572214   
7        109-4    1555589120              1967611.0    2760711405   
8        110-1    1658551290              2623824.0    2919131461   
9        110-2    1722679133              3171331.0    3092815189   
10       110-3    1828894565              3472190.0    3332311884   
11       110-4    1975118704              3309131.0    3725503455   
12       111-1    2104330654              3087714.0    3992676667   
13       111-2    2248318787              3318773.0    4345941335   
14       111-3    2411490251              3055178.0    4643301766   
15       111-4    2693836970              4760047.0    4964778878   
16       112-1    2833396892              6147668.0    5045844348   
17       112-2    2947233393              5611138.0    5149465045   
18       112-3    3132664367              5013050.0    5484556381   
19       112-4    3064474984              4382965.0    5532371215   
20       113-1    3051847989             10020132.0    5787891082   

    total_liabilities  shareholder_equity  \
0           443919183          1743517602   
1           684922144          1554421527   
2           546737449          1587497001   
3           642709606          1622095426   
4           666267165          1677028531   
5           727451140          1721213382   
6           847305839          1788266375   
7           910089406          1850621999   
8           978134499          1940996962   
9          1098844616          1993970573   
10         1253991574          2078320310   
11         1554770250          2170733205   
12         1671207913          2321468754   
13         1835479210          2510462125   
14         1890985607          2752316159   
15         2004290011          2960488867   
16         1952946750          3092897598   
17         1943996846          3205468199   
18         2111738274          3372818107   
19         2049108368          3483262847   
20         2122174828          3665716254   

    longterm_funds_to_fixed_assets_retio  
0                                   1.57  
1                                   1.36  
2                                   1.33  
3                                   1.20  
4                                   1.17  
5                                   1.15  
6                                   1.19  
7                                   1.19  
8                                   1.17  
9                                   1.16  
10                                  1.14  
11                                  1.10  
12                                  1.10  
13                                  1.12  
14                                  1.14  
15                                  1.10  
16                                  1.09  
17                                  1.09  
18                                  1.08  
19                                  1.14  
20                                  1.20

- The longterm_funds_to_fixed_assets_ratio for TSMC has fluctuated slightly between 1.10 and 1.20 over the past five years, indicating a balanced and consistent financial strategy.
- The slight decrease in the ratio in 2021 and 2022 reflects strategic investments in fixed assets, possibly for expansion or upgrading capabilities.
- The increase in the ratio in 2023 suggests a return to a slightly more conservative approach, increasing long-term funds relative to fixed assets, which enhances financial stability.
- The consistent ratios indicate that TSMC effectively manages its long-term funds to ensure sufficient coverage for its fixed assets, supporting sustained growth and operational resilience.

# Solvency (償債能力)

### Current ratio (流動比率)

Current ratio = current assets / current liabilities

流動比率 = 流動資產 / 流動負債

In [146]:
df_current_assets = get_acct_data_and_rename(df_balance, '流動資產合計', 'current_assets')
df_current_assets

year_season  current_assets
0        108-1       991324815
1        108-2      1010179338
2        108-3       849427436
3        108-4       822613914
4        109-1       802470040
5        109-2       860031580
6        109-3      1029673918
7        109-4      1092185308
8        110-1      1139786823
9        110-2      1241104023
10       110-3      1370639169
11       110-4      1607072907
12       111-1      1722237632
13       111-2      1905866396
14       111-3      2014232358
15       111-4      2052896744
16       112-1      1995727521
17       112-2      1959964207
18       112-3      2082477636
19       112-4      2194032910
20       113-1      2452767394

In [147]:
df_current_liabilities = get_acct_data_and_rename(df_balance, '流動負債合計', 'current_liabilities')
df_current_liabilities

year_season  current_liabilities
0        108-1            378267634
1        108-2            622256378
2        108-3            494781125
3        108-4            590735701
4        109-1            589468059
5        109-2            614795728
6        109-3            588241723
7        109-4            617151048
8        110-1            661648612
9        110-2            647748947
10       110-3            655621450
11       110-4            739503358
12       111-1            822867707
13       111-2            845240981
14       111-3            807431287
15       111-4            944226817
16       112-1            873089921
17       112-2            810829007
18       112-3            970034816
19       112-4            913583316
20       113-1           1026180079

In [148]:
df_assets_liabilities = pd.merge(df_current_assets, df_current_liabilities, how='inner', on='year_season')
df_assets_liabilities['current_ratio'] = (round(df_assets_liabilities['current_assets'] / df_assets_liabilities['current_liabilities'] *100, 2)).astype(str) + '%'
df_assets_liabilities

year_season  current_assets  current_liabilities current_ratio
0        108-1       991324815            378267634       262.07%
1        108-2      1010179338            622256378       162.34%
2        108-3       849427436            494781125       171.68%
3        108-4       822613914            590735701       139.25%
4        109-1       802470040            589468059       136.13%
5        109-2       860031580            614795728       139.89%
6        109-3      1029673918            588241723       175.04%
7        109-4      1092185308            617151048       176.97%
8        110-1      1139786823            661648612       172.26%
9        110-2      1241104023            647748947        191.6%
10       110-3      1370639169            655621450       209.06%
11       110-4      1607072907            739503358       217.32%
12       111-1      1722237632            822867707        209.3%
13       111-2      1905866396            845240981       225.48%
14       111-3      2014232358            807431287       249.46%
15       111-4      2052896744            944226817       217.42%
16       112-1      1995727521            873089921       228.58%
17       112-2      1959964207            810829007       241.72%
18       112-3      2082477636            970034816       214.68%
19       112-4      2194032910            913583316       240.16%
20       113-1      2452767394           1026180079       239.02%

- Current ratio is bigger than 100% which is healthy.
- The meaning of healthy is that the company has the ability to pay the liabilities.

流動比率大於100%為佳， 若小於100%須瞭解流動資產的變現性， 還有短期金融負債(短期借款、應附商業本票和一年內到期長期負債)佔流動負債的比例。 如果流動資產變現性很高，短期金融負債佔流動負債的比例低， 則低流動比率可接受；反之則償債能力不佳。 短期償債壓力越大，投資應特別注意。

In [149]:
# def for calculating short term liabilities

### Quick Ratio(速動比率)

速動比率 =（流動資產-存貨）/流動負債

In [150]:
# 存貨
df_inventories = get_acct_data_and_rename(df_balance, '存貨', 'inventories')
df_inventories

year_season  inventories
0        108-1    108682382
1        108-2    108231879
2        108-3     96685730
3        108-4     82981196
4        109-1     78277834
5        109-2     85788094
6        109-3    109509486
7        109-4    137353407
8        110-1    154567568
9        110-2    170438686
10       110-3    182233752
11       110-4    193102321
12       111-1    200118682
13       111-2    217438091
14       111-3    218335874
15       111-4    221149148
16       112-1    216067981
17       112-2    234332949
18       112-3    262090072
19       112-4    250997088
20       113-1    267123339

In [151]:
df_assets_liabilities = pd.merge(df_current_assets, df_current_liabilities, how='inner', on='year_season')
df_assets_liabilities_inventories = pd.merge(df_assets_liabilities, df_inventories, how='inner', on='year_season')
df_assets_liabilities_inventories['quick_ratio'] = round((df_assets_liabilities_inventories['current_assets'] - df_assets_liabilities_inventories['inventories'])/df_assets_liabilities_inventories['current_liabilities']*100, 2).astype(str) + '%'
df_assets_liabilities_inventories

year_season  current_assets  current_liabilities  inventories quick_ratio
0        108-1       991324815            378267634    108682382     233.34%
1        108-2      1010179338            622256378    108231879     144.95%
2        108-3       849427436            494781125     96685730     152.14%
3        108-4       822613914            590735701     82981196     125.21%
4        109-1       802470040            589468059     78277834     122.86%
5        109-2       860031580            614795728     85788094     125.94%
6        109-3      1029673918            588241723    109509486     156.43%
7        109-4      1092185308            617151048    137353407     154.72%
8        110-1      1139786823            661648612    154567568      148.9%
9        110-2      1241104023            647748947    170438686     165.29%
10       110-3      1370639169            655621450    182233752     181.26%
11       110-4      1607072907            739503358    193102321     191.21%
12       111-1      1722237632            822867707    200118682     184.98%
13       111-2      1905866396            845240981    217438091     199.76%
14       111-3      2014232358            807431287    218335874     222.42%
15       111-4      2052896744            944226817    221149148     193.99%
16       112-1      1995727521            873089921    216067981     203.83%
17       112-2      1959964207            810829007    234332949     212.82%
18       112-3      2082477636            970034816    262090072     187.66%
19       112-4      2194032910            913583316    250997088     212.68%
20       113-1      2452767394           1026180079    267123339     212.99%

# Operating ability(經營能力)

### Receivables Turnover Ratio(應收款項週轉率)

- 銷貨淨額: Revenue 
- 各期平均應收款項餘額: Average Accounts Receivable
- DSO = (Average Accounts Receivable) / (Revenue) * 365

- Receivables Turnover Ratio = (Revenue) / Average Accounts Receivable
- 應收款項週轉率 = 銷貨淨額 / 各期平均應收款項餘額

In [152]:
# 應收帳款淨額
df_receivables = get_acct_data_and_rename(df_balance, '應收帳款淨額', 'receivables')
df_receivables

year_season  receivables
0        108-1    106431149
1        108-2    115725802
2        108-3    144240857
3        108-4    138908589
4        109-1    145992744
5        109-2    148708055
6        109-3    161700332
7        109-4    145480272
8        110-1    167143336
9        110-2    177475493
10       110-3    186309080
11       110-4    197586109
12       111-1    212846891
13       111-2    220996537
14       111-3    260447134
15       111-4    229755887
16       112-1    147715125
17       112-2    190164242
18       112-3    222467653
19       112-4    201313914
20       113-1    201500334

In [153]:
# 銷貨淨額
df_net_revenue = get_acct_data_and_rename(df_profitloss, '營業毛利（毛損）淨額', 'net_revenue')
df_net_revenue

year_season  net_revenue
0        108-1     90357557
1        108-2    103730060
2        108-3    139412434
3     108-year    492701896
4        109-1    160776508
5        109-2    164623791
6        109-3    190493496
7     109-year    711130120
8        110-1    189839430
9        110-2    186196818
10       110-3    212746195
11    110-year    819537266
12       111-1    273203166
13       111-2    315467994
14       111-3    370498717
15    111-year   1348354806
16       112-1    286500392
17       112-2    260199847
18       112-3    296642464
19    112-year   1175110628
20       113-1    314505269
21       108-4    159201845
22       109-4    195236325
23       110-4    230754823
24       111-4    389184929
25       112-4    331767925

In [154]:
df_receivable_turnover = pd.merge(df_receivables, df_net_revenue, how='inner', on='year_season')
df_receivable_turnover['receivable_turnover_ratio'] = (df_receivable_turnover['receivables'] / df_receivable_turnover['net_revenue'])
df_receivable_turnover

year_season  receivables  net_revenue  receivable_turnover_ratio
0        108-1    106431149     90357557                   1.177889
1        108-2    115725802    103730060                   1.115644
2        108-3    144240857    139412434                   1.034634
3        108-4    138908589    159201845                   0.872531
4        109-1    145992744    160776508                   0.908048
5        109-2    148708055    164623791                   0.903321
6        109-3    161700332    190493496                   0.848850
7        109-4    145480272    195236325                   0.745150
8        110-1    167143336    189839430                   0.880446
9        110-2    177475493    186196818                   0.953161
10       110-3    186309080    212746195                   0.875734
11       110-4    197586109    230754823                   0.856260
12       111-1    212846891    273203166                   0.779079
13       111-2    220996537    315467994                   0.700536
14       111-3    260447134    370498717                   0.702964
15       111-4    229755887    389184929                   0.590351
16       112-1    147715125    286500392                   0.515584
17       112-2    190164242    260199847                   0.730839
18       112-3    222467653    296642464                   0.749952
19       112-4    201313914    331767925                   0.606791
20       113-1    201500334    314505269                   0.640690

### Inventory Turnover(存貨周轉率)

公司在某一段時間的營業成本與平均存貨餘額的比例，可以反應存貨的周轉速度，藉此看出存貨流動性、存貨資金占用量是否合理。

存貨周轉率 = 營業成本 / 平均存貨 = 營業成本 / ((期初存貨 + 期末存貨) / 2)

Inventroy Turnover = 

In [155]:
df_operation_cost = get_acct_data_and_rename(df_profitloss, '營業成本合計', 'operation_cost')
df_operation_cost

year_season  operation_cost
0        108-1       128352344
1        108-2       137325245
2        108-3       153613278
3     108-year       577286947
4        109-1       149813002
5        109-2       146005542
6        109-3       165945534
7     109-year       628108309
8        110-1       172635011
9        110-2       185910987
10       110-3       201821695
11    110-year       767877771
12       111-1       217872707
13       111-2       218672814
14       111-3       242644026
15    111-year       915536486
16       112-1       222132581
17       112-2       220641407
18       112-3       250090294
19    112-year       986625213
20       113-1       278138932
21       108-4       157996080
22       109-4       166344231
23       110-4       207510078
24       111-4       236346939
25       112-4       293760931

In [156]:
df_inventories = get_acct_data_and_rename(df_balance, '存貨', 'inventories')
df_inventories

year_season  inventories
0        108-1    108682382
1        108-2    108231879
2        108-3     96685730
3        108-4     82981196
4        109-1     78277834
5        109-2     85788094
6        109-3    109509486
7        109-4    137353407
8        110-1    154567568
9        110-2    170438686
10       110-3    182233752
11       110-4    193102321
12       111-1    200118682
13       111-2    217438091
14       111-3    218335874
15       111-4    221149148
16       112-1    216067981
17       112-2    234332949
18       112-3    262090072
19       112-4    250997088
20       113-1    267123339

In [157]:
df_inventories['avg_inventories'] = (df_inventories['inventories'] + df_inventories['inventories'].shift(1)) / 2
df_inventories['avg_inventories'] = np.where(df_inventories['avg_inventories'].isnull(), df_inventories['inventories'], df_inventories['avg_inventories'])
df_inventories

year_season  inventories  avg_inventories
0        108-1    108682382      108682382.0
1        108-2    108231879      108457130.5
2        108-3     96685730      102458804.5
3        108-4     82981196       89833463.0
4        109-1     78277834       80629515.0
5        109-2     85788094       82032964.0
6        109-3    109509486       97648790.0
7        109-4    137353407      123431446.5
8        110-1    154567568      145960487.5
9        110-2    170438686      162503127.0
10       110-3    182233752      176336219.0
11       110-4    193102321      187668036.5
12       111-1    200118682      196610501.5
13       111-2    217438091      208778386.5
14       111-3    218335874      217886982.5
15       111-4    221149148      219742511.0
16       112-1    216067981      218608564.5
17       112-2    234332949      225200465.0
18       112-3    262090072      248211510.5
19       112-4    250997088      256543580.0
20       113-1    267123339      259060213.5

In [158]:
df_inventories_turnover = pd.merge(df_inventories, df_operation_cost, how='inner', on='year_season')
df_inventories_turnover['inventories_turnover'] = round(df_inventories_turnover['operation_cost'] / df_inventories_turnover['avg_inventories'], 2)
# Days sales of invetory
df_inventories_turnover['DSI'] = round(365 / df_inventories_turnover['inventories_turnover'], 2)
df_inventories_turnover

year_season  inventories  avg_inventories  operation_cost  \
0        108-1    108682382      108682382.0       128352344   
1        108-2    108231879      108457130.5       137325245   
2        108-3     96685730      102458804.5       153613278   
3        108-4     82981196       89833463.0       157996080   
4        109-1     78277834       80629515.0       149813002   
5        109-2     85788094       82032964.0       146005542   
6        109-3    109509486       97648790.0       165945534   
7        109-4    137353407      123431446.5       166344231   
8        110-1    154567568      145960487.5       172635011   
9        110-2    170438686      162503127.0       185910987   
10       110-3    182233752      176336219.0       201821695   
11       110-4    193102321      187668036.5       207510078   
12       111-1    200118682      196610501.5       217872707   
13       111-2    217438091      208778386.5       218672814   
14       111-3    218335874      217886982.5       242644026   
15       111-4    221149148      219742511.0       236346939   
16       112-1    216067981      218608564.5       222132581   
17       112-2    234332949      225200465.0       220641407   
18       112-3    262090072      248211510.5       250090294   
19       112-4    250997088      256543580.0       293760931   
20       113-1    267123339      259060213.5       278138932   

    inventories_turnover     DSI  
0                   1.18  309.32  
1                   1.27  287.40  
2                   1.50  243.33  
3                   1.76  207.39  
4                   1.86  196.24  
5                   1.78  205.06  
6                   1.70  214.71  
7                   1.35  270.37  
8                   1.18  309.32  
9                   1.14  320.18  
10                  1.14  320.18  
11                  1.11  328.83  
12                  1.11  328.83  
13                  1.05  347.62  
14                  1.11  328.83  
15                  1.08  337.96  
16                  1.02  357.84  
17                  0.98  372.45  
18                  1.01  361.39  
19                  1.15  317.39  
20                  1.07  341.12

### Asset Turnover Ratio(資產周轉率)


- 資產周轉率 = 總營業收入(Total Revenue) ÷ 總資產(Total Asset)
- 總資產(Total Asset) = (期初資產+期末資產) ÷2



In [159]:
df_total_revenue = get_acct_data_and_rename(df_profitloss, '營業收入合計', 'total_revenue')
df_total_revenue

year_season  total_revenue
0        108-1      218704469
1        108-2      240998475
2        108-3      293045439
3     108-year     1069985448
4        109-1      310597183
5        109-2      310698367
6        109-3      356426204
7     109-year     1339254811
8        110-1      362410230
9        110-2      372145122
10       110-3      414670379
11    110-year     1587415037
12       111-1      491075873
13       111-2      534140808
14       111-3      613142743
15    111-year     2263891292
16       112-1      508632973
17       112-2      480841254
18       112-3      546732758
19    112-year     2161735841
20       113-1      592644201
21       108-4      317237065
22       109-4      361533057
23       110-4      438189306
24       111-4      625531868
25       112-4      625528856

In [160]:
df_total_assets = get_acct_data_and_rename(df_balance, '資產總額', 'total_assets')
df_total_assets['avg_assets'] = (df_total_assets['total_assets'] + df_total_assets['total_assets'].shift(1)) / 2
df_total_assets['avg_assets'] = np.where(df_total_assets['avg_assets'].isnull(), df_total_assets['total_assets'], df_total_assets['avg_assets'])
df_total_assets

year_season  total_assets    avg_assets
0        108-1    2187436785  2.187437e+09
1        108-2    2239343671  2.213390e+09
2        108-3    2134234450  2.186789e+09
3        108-4    2264805032  2.199520e+09
4        109-1    2343295696  2.304050e+09
5        109-2    2448664522  2.395980e+09
6        109-3    2635572214  2.542118e+09
7        109-4    2760711405  2.698142e+09
8        110-1    2919131461  2.839921e+09
9        110-2    3092815189  3.005973e+09
10       110-3    3332311884  3.212564e+09
11       110-4    3725503455  3.528908e+09
12       111-1    3992676667  3.859090e+09
13       111-2    4345941335  4.169309e+09
14       111-3    4643301766  4.494622e+09
15       111-4    4964778878  4.804040e+09
16       112-1    5045844348  5.005312e+09
17       112-2    5149465045  5.097655e+09
18       112-3    5484556381  5.317011e+09
19       112-4    5532371215  5.508464e+09
20       113-1    5787891082  5.660131e+09

In [161]:
(5.005312e+09 + 5.508464e+09)/2

5256888000.0

In [162]:
df_total_assets_turnover = pd.merge(df_total_revenue, df_total_assets[['year_season', 'avg_assets']],  how='inner', on='year_season')
df_total_assets_turnover['assets_turnover_ratio'] = round(df_total_assets_turnover['total_revenue'] / df_total_assets_turnover['avg_assets'], 2)
df_total_assets_turnover

year_season  total_revenue    avg_assets  assets_turnover_ratio
0        108-1      218704469  2.187437e+09                   0.10
1        108-2      240998475  2.213390e+09                   0.11
2        108-3      293045439  2.186789e+09                   0.13
3        109-1      310597183  2.304050e+09                   0.13
4        109-2      310698367  2.395980e+09                   0.13
5        109-3      356426204  2.542118e+09                   0.14
6        110-1      362410230  2.839921e+09                   0.13
7        110-2      372145122  3.005973e+09                   0.12
8        110-3      414670379  3.212564e+09                   0.13
9        111-1      491075873  3.859090e+09                   0.13
10       111-2      534140808  4.169309e+09                   0.13
11       111-3      613142743  4.494622e+09                   0.14
12       112-1      508632973  5.005312e+09                   0.10
13       112-2      480841254  5.097655e+09                   0.09
14       112-3      546732758  5.317011e+09                   0.10
15       113-1      592644201  5.660131e+09                   0.10
16       108-4      317237065  2.199520e+09                   0.14
17       109-4      361533057  2.698142e+09                   0.13
18       110-4      438189306  3.528908e+09                   0.12
19       111-4      625531868  4.804040e+09                   0.13
20       112-4      625528856  5.508464e+09                   0.11

### Fixed Assets Turnover Ratio, FAT(固定資產周轉率)

- 固定資產周轉率 = 營業收入淨額÷固定資產平均淨值
- 固定資產平均淨值 = (期初淨值+期末淨值)÷2

In [163]:
# 固定資產: 不動產、廠房及設備
df_fixed_assets = get_acct_data_and_rename(df_balance, '不動產、廠房及設備', 'fixed_assets')
df_fixed_assets

year_season  fixed_assets
0        108-1    1107651816
1        108-2    1142871184
2        108-3    1197955298
3        108-4    1352377405
4        109-1    1438215285
5        109-2    1493560950
6        109-3    1504438119
7        109-4    1555589120
8        110-1    1658551290
9        110-2    1722679133
10       110-3    1828894565
11       110-4    1975118704
12       111-1    2104330654
13       111-2    2248318787
14       111-3    2411490251
15       111-4    2693836970
16       112-1    2833396892
17       112-2    2947233393
18       112-3    3132664367
19       112-4    3064474984
20       113-1    3051847989

In [164]:
df_total_revenue = get_acct_data_and_rename(df_profitloss, '營業收入合計', 'total_revenue')
df_total_revenue

year_season  total_revenue
0        108-1      218704469
1        108-2      240998475
2        108-3      293045439
3     108-year     1069985448
4        109-1      310597183
5        109-2      310698367
6        109-3      356426204
7     109-year     1339254811
8        110-1      362410230
9        110-2      372145122
10       110-3      414670379
11    110-year     1587415037
12       111-1      491075873
13       111-2      534140808
14       111-3      613142743
15    111-year     2263891292
16       112-1      508632973
17       112-2      480841254
18       112-3      546732758
19    112-year     2161735841
20       113-1      592644201
21       108-4      317237065
22       109-4      361533057
23       110-4      438189306
24       111-4      625531868
25       112-4      625528856

In [165]:
df_net_revenue = get_acct_data_and_rename(df_profitloss, '營業毛利（毛損）淨額', 'net_revenue')
df_net_revenue

year_season  net_revenue
0        108-1     90357557
1        108-2    103730060
2        108-3    139412434
3     108-year    492701896
4        109-1    160776508
5        109-2    164623791
6        109-3    190493496
7     109-year    711130120
8        110-1    189839430
9        110-2    186196818
10       110-3    212746195
11    110-year    819537266
12       111-1    273203166
13       111-2    315467994
14       111-3    370498717
15    111-year   1348354806
16       112-1    286500392
17       112-2    260199847
18       112-3    296642464
19    112-year   1175110628
20       113-1    314505269
21       108-4    159201845
22       109-4    195236325
23       110-4    230754823
24       111-4    389184929
25       112-4    331767925

In [166]:
df_fixed_assets_turnover = pd.merge(df_total_revenue, df_fixed_assets,  how='inner', on='year_season')
df_fixed_assets_turnover['assets_turnover_ratio'] = round(df_fixed_assets_turnover['total_revenue'] / df_fixed_assets_turnover['fixed_assets'], 2)
df_fixed_assets_turnover

year_season  total_revenue  fixed_assets  assets_turnover_ratio
0        108-1      218704469    1107651816                   0.20
1        108-2      240998475    1142871184                   0.21
2        108-3      293045439    1197955298                   0.24
3        109-1      310597183    1438215285                   0.22
4        109-2      310698367    1493560950                   0.21
5        109-3      356426204    1504438119                   0.24
6        110-1      362410230    1658551290                   0.22
7        110-2      372145122    1722679133                   0.22
8        110-3      414670379    1828894565                   0.23
9        111-1      491075873    2104330654                   0.23
10       111-2      534140808    2248318787                   0.24
11       111-3      613142743    2411490251                   0.25
12       112-1      508632973    2833396892                   0.18
13       112-2      480841254    2947233393                   0.16
14       112-3      546732758    3132664367                   0.17
15       113-1      592644201    3051847989                   0.19
16       108-4      317237065    1352377405                   0.23
17       109-4      361533057    1555589120                   0.23
18       110-4      438189306    1975118704                   0.22
19       111-4      625531868    2693836970                   0.23
20       112-4      625528856    3064474984                   0.20

# Profitability(獲利能力)

資產報酬率(Return of Assets, ROA) = [稅後淨利 + 利息費用 X ( 1 - 稅率 )] / 平均資產總額

根據玩股網的公式
分子僅算稅後淨利

稅後損益: post-tax profit, PAT
稅後淨利 = 稅前淨利 - 所得稅
利息費用: Interest Expense

In [167]:
# 稅前淨利（淨損）	:
df_pretax_profit = get_acct_data_and_rename(df_profitloss, '稅前淨利（淨損）', 'pre_tax_profit')

# 所得稅費用（利益）合計
df_income_tax = get_acct_data_and_rename(df_profitloss, '所得稅費用（利益）合計', 'income_tax')

# 稅後淨利: 本期淨利（淨損）
df_posttax_profit = pd.merge(df_pretax_profit, df_income_tax, how='inner', on='year_season')
df_posttax_profit['posttax_profit'] = df_posttax_profit['pre_tax_profit'] - df_posttax_profit['income_tax']
df_posttax_profit = pd.merge(df_posttax_profit, df_total_assets, how='inner', on='year_season')
df_posttax_profit['year'] = df_posttax_profit['year_season'].apply(lambda x : x.split('-')[0])
df_posttax_profit['season'] = df_posttax_profit['year_season'].apply(lambda x : x.split('-')[1])
df_posttax_profit

year_season  pre_tax_profit  income_tax  posttax_profit  total_assets  \
0        108-1        68181652     6794342        61387310    2187436785   
1        108-2        80545440    13769589        66775851    2239343671   
2        108-3       112336271    11233817       101102454    2134234450   
3        109-1       132147178    15084285       117062893    2343295696   
4        109-2       136398638    15506124       120892514    2448664522   
5        109-3       155124003    17745852       137378151    2635572214   
6        110-1       155064389    15325640       139738749    2919131461   
7        110-2       149391075    14909359       134481716    3092815189   
8        110-3       173851639    17372485       156479154    3332311884   
9        111-1       226831696    23958322       202873374    3992676667   
10       111-2       265998281    28818029       237180252    4345941335   
11       111-3       316690867    35722460       280968407    4643301766   
12       112-1       244274931    37325895       206949036    5045844348   
13       112-2       214674909    32957903       181717006    5149465045   
14       112-3       241940495    31145221       210795274    5484556381   
15       113-1       266543204    41321941       225221263    5787891082   
16       108-4       128781973    12703779       116078194    2264805032   
17       109-4       161107361    18282837       142824524    2760711405   
18       110-4       184819211    18445696       166373515    3725503455   
19       111-4       334669874    38791392       295878482    4964778878   
20       112-4       278280989    39974788       238306201    5532371215   

      avg_assets year season  
0   2.187437e+09  108      1  
1   2.213390e+09  108      2  
2   2.186789e+09  108      3  
3   2.304050e+09  109      1  
4   2.395980e+09  109      2  
5   2.542118e+09  109      3  
6   2.839921e+09  110      1  
7   3.005973e+09  110      2  
8   3.212564e+09  110      3  
9   3.859090e+09  111      1  
10  4.169309e+09  111      2  
11  4.494622e+09  111      3  
12  5.005312e+09  112      1  
13  5.097655e+09  112      2  
14  5.317011e+09  112      3  
15  5.660131e+09  113      1  
16  2.199520e+09  108      4  
17  2.698142e+09  109      4  
18  3.528908e+09  110      4  
19  4.804040e+09  111      4  
20  5.508464e+09  112      4

In [168]:
df_112 = df_posttax_profit[df_posttax_profit['year_season'].str.startswith('112')]
avg_assets_total_112 = df_112['avg_assets'].mean()
posttax_profit_112 = df_112['posttax_profit'].sum()
df_112

year_season  pre_tax_profit  income_tax  posttax_profit  total_assets  \
12       112-1       244274931    37325895       206949036    5045844348   
13       112-2       214674909    32957903       181717006    5149465045   
14       112-3       241940495    31145221       210795274    5484556381   
20       112-4       278280989    39974788       238306201    5532371215   

      avg_assets year season  
12  5.005312e+09  112      1  
13  5.097655e+09  112      2  
14  5.317011e+09  112      3  
20  5.508464e+09  112      4

In [169]:
# 當季資產報酬率 = 當季poasttax_profit / (前一期total_assets + 當期total_assets)/2
238306201/((5532371215+5484556381)/2)

0.04326182575376526

In [175]:
df_each_year.sort_values('season', ascending=True)

year_season  pre_tax_profit  income_tax  posttax_profit  total_assests  \
6        110-1       155064389    15325640       139738749     2919131461   
7        110-2       149391075    14909359       134481716     3092815189   
8        110-3       173851639    17372485       156479154     3332311884   
18       110-4       184819211    18445696       166373515     3725503455   

      avg_assets year season  
6   2.839921e+09  110      1  
7   3.005973e+09  110      2  
8   3.212564e+09  110      3  
18  3.528908e+09  110      4

In [215]:
year_list = sorted({int(i[0]) for i in df_posttax_profit.year_season.str.split('-')})
year_list

[108, 109, 110, 111, 112, 113]

In [220]:
df_ROA = pd.DataFrame()
year_list = sorted({str(i[0]) for i in df_posttax_profit.year_season.str.split('-')})
for year in year_list[0:3]:
    df_each_year = df_posttax_profit[df_posttax_profit['year'] == year]
    season_list = df_posttax_profit[df_posttax_profit['year'] == year].season.to_list()
    if len(season_list) > 1:
        # calculate the return of assets(ROA) in each season in the year
        df_each_year = df_each_year.sort_values('season', ascending=True)
        for _, row in df_each_year.iterrows():
            row['ROA'] = row['posttax_profit'] / row['avg_assets']
            
        df_ROA = pd.concat([df_ROA, df_each_year], axis=0)

In [224]:
row

year_season              110-4
pre_tax_profit       184819211
income_tax            18445696
posttax_profit       166373515
total_assets        3725503455
avg_assets        3528907669.5
year                       110
season                       4
ROA                   0.047146
Name: 18, dtype: object

In [122]:
season_list = df_posttax_profit[df_posttax_profit['year'] == year].season.to_list()
len(season_list)

4

In [115]:
for year in year_list:
    print(year)

108
109
110
111
112
113


In [42]:
# dominator: the avg of previous season total assets and target season
def posttax_profit(df):
    # get the years in the data
    
    
    return 0